File: Exercises/101-109 - Small Problems/medium_2/sum_square_square_sum.rb
# Write a method that computes the difference between the square of the sum of the first n positive integers
# and the sum of the squares of the first n positive integers.

# Examples:

# ```
# sum_square_difference(3) == 22 # -> (1 + 2 + 3)**2 - (1**2 + 2**2 + 3**2)
# sum_square_difference(10) == 2640
# sum_square_difference(1) == 0
# sum_square_difference(100) == 25164150
# ```

def sum_square_difference(n)
  sum_square = (1..n).reduce(:+)**2
  square_sum = (1..n).reduce { |total, num| total += num**2 }

  sum_square - square_sum
end

puts sum_square_difference(3) == 22
puts sum_square_difference(10) == 2640
puts sum_square_difference(1) == 0
puts sum_square_difference(100) == 25164150

File: Exercises/210 - Small Problems/medium-2/sum-square-square-sum.js
/*
Write a function that computes the difference between the square of the sum of the first n positive integers and the sum of the squares of the first n positive integers.

Examples:

sumSquareDifference(3)    // 22
// -> (1 + 2 + 3)**2 - (1**2 + 2**2 + 3**2)
sumSquareDifference(10)   // 2640
sumSquareDifference(1)    // 0
sumSquareDifference(100)  // 25164150
*/

// loop from 1 to num inclusively.
// at each iteration
//  - add i to the sum
//  - add Math.pow(i, 2) to the square
// return Math.pow(sum, 2) - square

function sumSquareDifference(num) {
  var sum = 0;
  var square = 0;
  var i;

  for (i = 1; i <= num; i += 1) {
    sum += i;
    square += Math.pow(i, 2);
  }

  return Math.pow(sum, 2) - square;
}

console.log(sumSquareDifference(3));
console.log(sumSquareDifference(10));
console.log(sumSquareDifference(1));
console.log(sumSquareDifference(100));